sycl/doc/extensions/experimental/sycl_ext_private_alloca.asciidoc

```diff
@@ -0,0 +1,266 @@
+sequences of objects. However, these constructs have a significant restriction:
+the number of elements must be known at compile time. In host-code context,
+users can make use of dynamic memory allocations, e.g., `std::vector`, but this
+is not the case in SYCL device code.
+
+SYCL specialization constants (SYCL Section 4.9.5.) can be used to represent
+constants whose values can be set dynamically during the execution of a SYCL
+application, but that will not change when a SYCL kernel function is
+invoked. This way, specialization constants could be used to implement SYCL
+private arrays whose size is given during the execution of the SYCL
+application. There is no possible way of implementing this using `std::array`,
+as the size of such containers must be known at compile time, so we propose to
+define a new `private_alloca` function whose size is specified using SYCL
+specialization constants.
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_PRIVATE_ALLOCA` to one of the values defined in the table
+below. Applications can test for the existence of this macro to determine if the
+implementation supports this feature, or applications can test the macro's value
+to determine which of the extension's features the implementation supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+=== New aspect for specialization constant-length allocations
+
+This extension adds a new device aspect:
+
+[source,c++]
+----
+namespace sycl {
+enum class aspect : /*unspecified*/ {
+  ext_oneapi_private_alloca
+};
+} // namespace sycl
+----
+
+The `ext_oneapi_private_alloca` aspect indicates that the device is capable of
+using the `private_alloca` API defined in the following sections.
+
+=== The `private_alloca` API
+
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+template <typename ElementType, auto &SpecName,
+          access::decorated DecorateAddress>
+private_ptr<ElementType, DecorateAddress>
+private_alloca(kernel_handler &kh);
+
+template <typename ElementType, std::size_t Alignment, auto &SpecName,
+          access::decorated DecorateAddress>
+private_ptr<ElementType, DecorateAddress>
+aligned_private_alloca(kernel_handler &kh);
+} // namespace sycl::ext::oneapi::experimental
+----
+
+_Mandates_: `ElementType` must be a cv-unqualified trivial type and `SpecName`
+must be a reference to a specialization constant of integral `value_type`. In
+the case of `aligned_private_alloca`, `Alignment` must be an alignment value and
+must be a positive multiple of `alignof(ElementType)`. If `Alignment` is an
+extended alignment, it must be supported by the implementation.
```

Comment:
Just for my understanding ... isn't `aligned_private_alloca` only useful if `Alignment` is an extended alignment?  The last sentence makes is sound like the user may request some alignment that is not an extended alignment, but why would you do that?  If you only want the fundamental alignment, don't you get that automatically with `private_alloca`?

Reply:
*Fundamental alignment* means an alignment value the implementation supports in all contexts (less than or equal to `alignof(std::max_align_t)`). *Extended alignment* means an alignment value not supported in all contexts by the implementation (greater than `alignof(std::max_align_t)`). Note *fundamental alignment* is different to a type's *alignment requirements*, meaning the value returned by `alignof(T)`. [See](https://eel.is/c++draft/basic.align).

```c++
// Aligned to 'float' alignment requirement:
auto ptr0 = std::private_alloca<float, size, decorated>(h);
// Aligned to fundamental alignment:
auto ptr1 = std::aligned_private_alloca<float, alignof(double), size, decorated>(h);
// Aligned to extended alignment:
auto ptr2 = std::aligned_private_alloca<float, 2 * alignof(std::max_align_t), size, decorated>(h);
```